enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+	//
+	// The lock MUST be held until ALL modifications have been effected.
+	//
+	// It is run with RepeatableRead isolation, so it's effectively snapshotting the data at the start of the tx.
+	//
+	// This is a read-only tx, so returning an error (i.e. causing a rollback) has no impact.
+	return c.store.InTx(func(db database.Store) error {
+		start := c.clock.Now()
+
+		// Try to acquire the lock. If we can't get it, another replica is handling reconciliation.
+		acquired, err := db.TryAcquireLock(ctx, database.LockIDReconcilePrebuilds)
+		if err != nil {
+			// This is a real database error, not just lock contention
+			logger.Error(ctx, "failed to acquire reconciliation lock due to database error", slog.Error(err))
+			return err
+		}
+		if !acquired {
+			// Normal case: another replica has the lock
+			return nil
+		}
+
+		logger.Debug(ctx, "acquired top-level reconciliation lock", slog.F("acquire_wait_secs", fmt.Sprintf("%.4f", c.clock.Since(start).Seconds())))
+
+		return fn(ctx, db)
+	}, &database.TxOptions{
+		Isolation:    sql.LevelRepeatableRead,
+		ReadOnly:     true,
+		TxIdentifier: "prebuilds",
+	})
+}
+
+func (c *StoreReconciler) createPrebuild(ctx context.Context, prebuildID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
+	name, err := prebuilds.GenerateName()
+	if err != nil {
+		return xerrors.Errorf("failed to generate unique prebuild ID: %w", err)
+	}
+
+	return c.store.InTx(func(db database.Store) error {
+		template, err := db.GetTemplateByID(ctx, templateID)
+		if err != nil {
+			return xerrors.Errorf("failed to get template: %w", err)
+		}
+
+		now := c.clock.Now()
+
+		minimumWorkspace, err := db.InsertWorkspace(ctx, database.InsertWorkspaceParams{
+			ID:                prebuildID,
+			CreatedAt:         now,
+			UpdatedAt:         now,
+			OwnerID:           prebuilds.SystemUserID,
+			OrganizationID:    template.OrganizationID,
+			TemplateID:        template.ID,
+			Name:              name,
+			LastUsedAt:        c.clock.Now(),
+			AutomaticUpdates:  database.AutomaticUpdatesNever,
+			AutostartSchedule: sql.NullString{},
+			Ttl:               sql.NullInt64{},
+			NextStartAt:       sql.NullTime{},
+		})
+		if err != nil {
+			return xerrors.Errorf("insert workspace: %w", err)
+		}
+
+		// We have to refetch the workspace for the joined in fields.
+		workspace, err := db.GetWorkspaceByID(ctx, minimumWorkspace.ID)
+		if err != nil {
+			return xerrors.Errorf("get workspace by ID: %w", err)
+		}
+
+		c.logger.Info(ctx, "attempting to create prebuild", slog.F("name", name),
+			slog.F("workspace_id", prebuildID.String()), slog.F("preset_id", presetID.String()))
+
+		return c.provision(ctx, db, prebuildID, template, presetID, database.WorkspaceTransitionStart, workspace)
+	}, &database.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  false,
+	})
+}
+
+func (c *StoreReconciler) deletePrebuild(ctx context.Context, prebuildID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
```

Comment:
Please try to stick to consistent terminology: prebuiltWorkspace for a workspace, prebuild for a build.
```suggestion
func (c *StoreReconciler) deletePrebuild(ctx context.Context, prebuiltWorkspaceID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
```

Reply:
I fixed it in a few places.
I'll try to do it, but the issue: this is already all over the prototype source code.